feat: Add Feast Operator RBAC example with Kubernetes Authentication …

redhatHameed opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#code-style-and-linting
2. Run unit tests and ensure that they are passing: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#unit-tests
3. If your change introduces any API changes, make sure to update the integration tests here: https://github.com/feast-dev/feast/tree/master/sdk/python/tests
4. Make sure documentation is updated for your PR!
5. Make sure your commits are signed: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#signing-off-commits
6. Make sure your PR title follows conventional commits (e.g. fix: [Description of ...], feat: [Description of ...], chore: [Description of ...], refactor: [Description of ...])

-->

# What this PR does / why we need it:
<!--
Outline what you're doing
-->
 Adding Feast Operator RBAC example with Kubernetes Authentication

# Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->


# Misc
<!--
Feel free to leave additional thoughts or tag people as you see fit
-->

redhatHameed commented:
@tchughesiv @franciscojavierarceo @dmartinol  when you get a chance, could you please provide review/feedback on this PR? Thanks!

redhatHameed commented on examples/operator-rbac/README.md:
```diff
@@ -0,0 +1,6 @@
+# Install and run a Feature Store on Kubernetes with the Feast Operator
```
done

accorvin commented:
@redhatHameed one thing that's not obvious to me is how the test.py script makes use of the service account that is being used to connect to the feature store. I'm assuming it's automatically using some environment variable somewhere or something like that, but I'm not sure. Can we document that so it's more explicit?

I expect users will want to be able to connect to a feature store instance from a script that isn't running inside of a kubernetes pod, and I'm assuming doing so would require them to specify the token somewhere, so we should make it clear how they can do so.

ntkathole commented on examples/operator-rbac/k8s-rbac.ipynb:
Install the Feast Operator section has quay image path, should it be dockerhub for now ? Also, `quay.io/feastdev-ci` or `quay.io/feastdev` ?

redhatHameed replied:
@ntkathole Thanks right, this will be updated with quay once this [PR](https://github.com/feast-dev/feast/pull/5047) merged

redhatHameed commented:
@accorvin can you take another look of this PR, I have updated the notebook to use local client instead of the pod. Thanks

franciscojavierarceo reviewed (commented):
I would contextualize some of the documentation that this is mostly relevant if you're an MLOps Engineer or Cluster Admin trying to add Feast to your k8s cluster.

redhatHameed commented:
> I would contextualize some of the documentation that this is mostly relevant if you're an MLOps Engineer or Cluster Admin trying to add Feast to your k8s cluster.

@franciscojavierarceo Thanks yes make sense, 

I have separated notebooks for setup operator and RBAC - > `1-setup-operator-rbac.ipynb` which need to perform Admin or MLOps Engineer, 

and client notebook -> `2-client.ipynb`  for specific to developer

franciscojavierarceo commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+    "# Feast Operator with RBAC Configuration\n",
+    "## Objective\n",
+    "\n",
+    "This demo provides a reference implementation of a runbook on how to enable Role-Based Access Control (RBAC) for Feast using the Feast Operator with the Kubernetes authentication type.\n",
```
```suggestion
    "This demo provides a reference implementation of a runbook on how to enable Role-Based Access Control (RBAC) for Feast using the Feast Operator with the Kubernetes authentication type. This serves as useful reference material for a cluster admin / MLOps engineer.\n",
```

franciscojavierarceo commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+   "source": [
+    "## Feast Client with RBAC\n",
+    "### RBAC Kubernetes Authentication\n",
+    "This Feast **Role-Based Access Control (RBAC)** in Kubernetes support authentication  **inside a Kubernetes pod** and **outside a pod** when running a local script.\n",
```
```suggestion
    "Feast **Role-Based Access Control (RBAC)** in Kubernetes supports authentication  **inside a Kubernetes pod** and **outside a pod** when running a local script.\n",
```

franciscojavierarceo commented on examples/operator-rbac/README.md:
```diff
@@ -0,0 +1,6 @@
+# Running the Feast RBAC example on Kubernetes using the Feast Operator.
+
+1. [1-setup-operator-rbac.ipynb](1-setup-operator-rbac.ipynb) will guide you through how to setup Role-Based Access Control (RBAC) for Feast using [Feast Operator](../../infra/feast-operator/)  with the Kubernetes Authentication type. This Feast Admin Step required to setup operator and Feast RBAC on K8s environment. 
```
```suggestion
1. [1-setup-operator-rbac.ipynb](1-setup-operator-rbac.ipynb) will guide you through how to setup Role-Based Access Control (RBAC) for Feast using the [Feast Operator](../../infra/feast-operator/)  and Kubernetes Authentication. This Feast Admin Step requires you to setup the operator and Feast RBAC on K8s. 
```

franciscojavierarceo commented on examples/operator-rbac/README.md:
```diff
@@ -0,0 +1,6 @@
+# Running the Feast RBAC example on Kubernetes using the Feast Operator.
+
+1. [1-setup-operator-rbac.ipynb](1-setup-operator-rbac.ipynb) will guide you through how to setup Role-Based Access Control (RBAC) for Feast using [Feast Operator](../../infra/feast-operator/)  with the Kubernetes Authentication type. This Feast Admin Step required to setup operator and Feast RBAC on K8s environment. 
+2. [2-client.ipynb](2-client.ipynb) Validate with RBAC with client example using different test cases with service account token locally.
```
```suggestion
2. [2-client.ipynb](2-client.ipynb) Validate the RBAC with the client example using different test cases using a service account token locally.
```

franciscojavierarceo commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+   },
+   "cell_type": "code",
+   "source": [
+    "!echo \"Running Feast RBAC test for Read only User...\"\n",
```
```suggestion
    "!echo \"Running the Feast RBAC test for a Read only User...\"\n",
```

franciscojavierarceo commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "**Step 2:Test function  features from offline, online and materialize_incremental etc**",
```
```suggestion
   "source": "**Step 2:Test misc functions from offline, online, materialize_incremental, and others**",
```

franciscojavierarceo commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+   "cell_type": "markdown",
+   "source": [
+    "## Feast Admin Steps:\n",
+    "Feast Admin or MLOps Engineers may require Kubernetes Cluster Admin roles when working with OpenShift or Kubernetes clusters. Below is the list of steps Required to perform setting up Feast RBAC with Operator by Admin or MLOps Engineers.\n",
```
```suggestion
    "Feast Admins or MLOps Engineers may require Kubernetes Cluster Admin roles when working with OpenShift or Kubernetes clusters. Below is the list of steps Required to set up Feast RBAC with the Operator by an Admin or MLOps Engineer.\n",
```

franciscojavierarceo reviewed (commented):
left some feedback.

dmartinol commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+    "* `Registry Server`: Handles metadata storage for feature definitions.\n",
+    "* `Online Store Server`: Uses the `Registry Server` to query metadata and is responsible for low-latency serving of features.\n",
+    "* `Offline Store Server`: Uses the `Registry Server` to query metadata and provides access to batch data for historical feature retrieval.\n",
+    "* `Kubernetes` Authentication types for RBAC Configuration for Feast resources.\n",
```
The last 2 bullets doesn't seem to be related to "distributed topology of Feast services".
The last one in particular, it looks like it's an objective of the notebook rather any architectural component.

redhatHameed replied:
make sense, updated

dmartinol commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "## Next Run Client notebook -> 2-client.ipynb"
```
can't this be a link instead?

dmartinol commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "## Configure the RBAC Permissions\n",
```
This section code be a bit more descriptive on the intentions, to clarify what the code is doing

redhatHameed replied:
added more details and comments in code as well to make it clear.

dmartinol commented on examples/operator-rbac/1-setup-operator-rbac.ipynb:
```diff
@@ -0,0 +1,753 @@
+     "start_time": "2025-03-05T18:50:19.362018Z"
+    }
+   },
+   "source": [
```
The same could be demonstrated by connecting a client to the registry server, w/o using the `kubectl exec` command. e.g., probably in the client notebook.
It probably depends on who's the target person we want to instruct with this notebook.

redhatHameed replied:
This is for Feast Admin part for setting up permission and then to validate applied permissions. Yup can be done with client and then connecting registry also I believe using `CLI`, if required.

dmartinol commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+    "- Developer just need create the binding with role and service account accordingly.\n",
+    "- Code Reference:  \n",
+    "[Feast Kubernetes Auth Client Manager (Pod Token Usage)](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/permissions/client/kubernetes_auth_client_manager.py#L15) \n",
+    "- See the example will use service account from pod [Example](https://github.com/feast-dev/feast/blob/master/examples/rbac-remote/client/k8s/)\n",
```
This sentence seems confusing, any typo?

redhatHameed replied:
updated to make it more clear

dmartinol commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+    "Feast will use this token for authentication.\n",
+    "\n",
+    "Reference:  \n",
+    "[Feast Authentication via `LOCAL_K8S_TOKEN`](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/permissions/client/kubernetes_auth_client_manager.py#L50)"
```
I would link to the user guide instead, there's a reference here:
https://docs.feast.dev/master/getting-started/components/authz_manager#kubernetes-rbac-authorization

redhatHameed replied:
Thanks for this, update the link

dmartinol commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "**The Operator creates the client ConfigMap containing the feature_store.yaml. We can retrieve it and port froward to local**",
```
what about `The Operator creates the ConfigMap containing the feature_store.yaml for a client application`

redhatHameed replied:
can you explain this more,  this required `port-forward` to locally and feature_repo required to point feature-server to local accordingly.

dmartinol commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "**Generating training data. The following test functions were copied from the `test_workflow.py` template but we added `try` blocks to print only \n",
```
all bold?

dmartinol commented on examples/operator-rbac/2-client.ipynb:
```diff
@@ -0,0 +1,918 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "An occurred while performing materialize incremental: \n",
```
There's no evidence that this is a permission error, is there another issue on the client side that ew can attach to the notebook?

redhatHameed replied:
remote materialization is missing feature, so removed from test script. added `store.apply(feature_views)` to validate write permission.

redhatHameed commented:
@franciscojavierarceo can we merge this so it can be part of today release if we are planning for that. Thanks

tchughesiv commented:
@redhatHameed while i understand the addition of workstation communication example ... where did the pod to pod (admin vs client) example go? shouldn't we have those as well?

tchughesiv commented:
i fear this is too complex ... maybe we pivot to demoing remote registry access here instead? much of the local workstation stuff would do better as a separate example, unrelated to RBAC.

tchughesiv commented:
I have an outstanding task to add an additional example around using the feature-server endpoints of an operator deployed feature store. i may play around with some of these rbac docs at the same time. i think this example would benefit from a remote registry use-case ... maybe even using the feature-server endpoints directly

redhatHameed commented:
> @redhatHameed while i understand the addition of local workstation client example ... where did the pod to pod (admin vs client) example go? shouldn't we have those as well? after all, restricting authorization from pods within the cluster is probably the more common use case here. its also the cleaner way to demonstrate how RBAC works in feast.
> 
> Can you add that part of the example back but keep the local workstation bit as secondary example as well?

@tchughesiv thanks for pointing that, I have updated and added that example in this PR #5147 which contain now client example using local environment and from the pod.
